### PR TITLE
Create Message model

### LINKED ISSUE #34: Need to add Message entity
Need to add:
1. Message model
2. Don't forget to add test for Message model

### PR DESCRIPTION
**Pull request by IdanZimi**
Create Message model

closes #34

### CODE REVIEW

**Comment by IdanZimi** [comment]
On this pull request I work with @OhadSeg.

**Comment by IdanZimi** [comment]
Anyone knows by chance why the checks are not running? @yftacherzog  @Yuval-Vino

**Comment by yftacherzog** [comment]
> Anyone knows by chance why the checks are not running? @yftacherzog  @Yuval-Vino 

Don't know but try rebasing on upstream/main.

**Comment by OhadSeg** [comment]
Hey @yftacherzog we tried  rebasing on upstream/main, but we have conflicts with one of our commits.

**Comment by yftacherzog** [comment]
> Hey @yftacherzog we tried  rebasing on upstream/main, but we have conflicts with one of our commits.

Are you having issues resolving them?

**Comment by OhadSeg** [comment]
@yftacherzog Working on it, it conflicts with our stub file for one of the dependency entities.

File: player/migrations/0001_initial.py [hunk 1]
```diff
@@ -1,22 +1,32 @@
+<<<<<<< HEAD
```
Comment (yftacherzog) [inline player/migrations/0001_initial.py:1]: You still have conflict markers that were missed during conflicts resolution.

Comment (yftacherzog) [reply player/migrations/0001_initial.py:1]: I'd suggest redoing the rebase and accepting all changes coming from upstream/main.

**Comment by IdanZimi** [comment]
@yftacherzog yes, were having issues with one of the migration files

File: player/migrations/0001_initial.py (hunk 1 above)
Comment (yftacherzog) [reply player/migrations/0001_initial.py:1]: Ping me tomorrow and we'll try to figure this out.

File: meetballs/settings.py [hunk 2]
```diff
@@ -38,7 +38,10 @@
     'django.contrib.messages',
     'django.contrib.staticfiles',
     'meet_balls_app.apps.MeetBallsAppConfig',
-    'player.apps.PlayerConfig'
+    'player.apps.PlayerConfig',
+    'message.apps.MessageConfig',
+    'game_event.apps.GameEventConfig',
+    'player.apps.PlayerConfig',
```
Comment (Danielsio) [inline meetballs/settings.py:44]: notice the duplicate lines, probably an outcome of a rebase.
 'player.apps.PlayerConfig',
 'message.apps.MessageConfig',

Comment (IdanZimi) [reply meetballs/settings.py]: amzing! thanks @Danielsio

**Comment by OhadSeg** [comment]
@yftacherzog @Yuval-Vino Could you review our PR ?

**Comment by yftacherzog** [comment]
> @yftacherzog @Yuval-Vino Could you review our PR ?

The tests are not running. You'd need to rebase (keeping your stubs) on upstream/main in order for  the tests to run. See that you address the the conflict in settings.py

File: message/models.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+
+class Message(models.Model):
+    user_id = models.ForeignKey(Player, on_delete=models.CASCADE)
+    game_event_id = models.ForeignKey(GameEvent, on_delete=models.CASCADE)
+    time_sent = models.DateTimeField(auto_now_add=True)
+    message_text = models.CharField(max_length=255)
+
+    class Meta:
+        unique_together = ('game_event_id', 'user_id')
```
Comment (Yuval-Vino) [inline message/models.py:15]: What is the purpose of this?
Does it imply that every user is restricted to sending just one message per event?

File: message/models.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+from game_event.models import GameEvent
+
+# Create your models here.
+
+
+class Message(models.Model):
+    user_id = models.ForeignKey(Player, on_delete=models.CASCADE)
+    game_event_id = models.ForeignKey(GameEvent, on_delete=models.CASCADE)
+    time_sent = models.DateTimeField(auto_now_add=True)
+    message_text = models.CharField(max_length=255)
```
Comment (Yuval-Vino) [inline message/models.py:12]: Do we really need to name it `message_text `? Since it is already a field in the Message class, it seems obvious that we are referring to the message itself. what do you think?

File: message/tests.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        updated_message = Message.objects.get(pk=message.pk)
+        assert updated_message.message_text == "test message?!"
+
+    def test_delete_message(self, saved_message):
+        message = saved_message
+        message.delete()
+        with pytest.raises(Message.DoesNotExist):
+            Message.objects.get(pk=saved_message.pk)
+
+    def test_delete_game_event_deletes_message(self, saved_message):
```
Comment (Yuval-Vino) [inline message/tests.py:62]: you can do the same test with delete user

**Comment by OhadSeg** [comment]
> > @yftacherzog @Yuval-Vino Could you review our PR ?
> 
> The tests are not running. You'd need to rebase (keeping your stubs) on upstream/main in order for the tests to run. See that you address the the conflict in settings.py

@yftacherzog we fixed it, can you check again please ?

File: message/tests.py (hunk 5 above)
Comment (OhadSeg) [reply message/tests.py:62]: @Yuval-Vino We are already in confusion if when deleting a user we also need to delete his message. wdyt ?

File: message/models.py (hunk 3 above)
Comment (Danielsio) [reply message/models.py:15]: Well there is the dateTime field there, so each messeage have unique dateTime, i am guessing that you can't send 2 messeages quick enough to make both have the same dateTime object, so it could work as it is right now.

File: message/models.py (hunk 4 above)
Comment (OhadSeg) [reply message/models.py:12]: @Yuval-Vino You're right here it is stated that it is a 'charField', but in other places in our code, we are accessing this attribute, if we omit the "_text" part of this attribute name, it could be a bit confusing just to try access message attribute in a  message entity. Wdyt ?

File: message/tests.py (hunk 5 above)
Comment (yftacherzog) [reply message/tests.py:62]: It's up to you. Currently, the model says it should be deleted (as they are cascaded on delete), so I'd say implement according to the model you have.

File: message/models.py (hunk 4 above)
Comment (Yuval-Vino) [reply message/models.py:12]: I was actually thinking of removing the message part from the field and just calling it text so you can refer to it as Message.text

File: message/models.py (hunk 3 above)
Comment (Yuval-Vino) [reply message/models.py:15]: the unique together function is for sets of field names that, taken together, must be unique.
have nothing to do with 2 foreign keys.
see more about unique-together: https://docs.djangoproject.com/en/3.1/ref/models/options/#unique-together

File: message/tests.py (hunk 5 above)
Comment (OhadSeg) [reply message/tests.py:62]: @yftacherzog @Yuval-Vino we add the relevant test for this case in accordance with what our model says.

File: message/models.py (hunk 4 above)
Comment (OhadSeg) [reply message/models.py:12]: @Yuval-Vino Seems like a good idea, working on it.

File: message/models.py (hunk 3 above)
Comment (OhadSeg) [reply message/models.py]: @Yuval-Vino we removed the unique together function from our model file.

File: message/tests.py
Comment (amitKaplansky) [inline message/tests.py:1]: Do you think  there's a need for a  test that checks a certain message isn't empty? @IdanZimi @OhadSeg

Comment (Danielsio) [reply message/tests.py:1]: if you do add a test case for that , you need to add the min_length attribute to the text field for the test to pass: 

text = models.CharField(max_length=255, min_length=1)

Comment (OhadSeg) [reply message/tests.py:1]: @amitKaplansky  we went with your comments, and added a test for an empty message, a long with updating our model, with max_length=255, validators=[MinLengthValidator(1)] values.
@Danielsio As we experienced and saw other cases as well, it is not possible to use both max_length and min_length, and that is why we use the MinLengthValidator.  
@Yuval-Vino @yftacherzog @Yuval-Vino Add test for empty message with changes in our model file.

File: message/admin.py [hunk 6]
```diff
@@ -0,0 +1,6 @@
+from django.contrib import admin
+from .models import Message
+
+admin.site.register(Message)
+
+# Register your models here.
```
Comment (yftacherzog) [inline message/admin.py:6]: delete this comment

File: message/models.py [hunk 7]
```diff
@@ -0,0 +1,13 @@
+from django.db import models
+from player.models import Player
+from game_event.models import GameEvent
+from django.core.validators import MinLengthValidator
+
+# Create your models here.
```
Comment (yftacherzog) [inline message/models.py:6]: delete this comment

File: message/admin.py (hunk 6 above)
Comment (IdanZimi) [reply message/admin.py:6]: Deleted @yftacherzog

File: message/tests.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        message = Message.objects.get(pk=saved_message.pk)
+        assert message == saved_message
+
+    def test_update_message(self, saved_message):
+        message = saved_message
+        assert message.text != "test message?!"
+        message.text = "test message?!"
+        message.save()
+        updated_message = Message.objects.get(pk=message.pk)
+        assert updated_message.text == "test message?!"
```
Comment (yftacherzog) [inline message/tests.py:54]: Making it more readable and maintainable:
```suggestion
    def test_update_message(self, saved_message):
        new_text = "test message?!"
        message = saved_message
        assert message.text != new_text
        message.text = new_text
        message.save()
        updated_message = Message.objects.get(pk=message.pk)
        assert updated_message.text == new_text
```

**Comment by yftacherzog** [comment]
@IdanZimi @OhadSeg lgtm. Good job. I'll do another sweep after the event PR is merged and you rebase on top of main.

File: message/tests.py (hunk 8 above)
Comment (IdanZimi) [reply message/tests.py]: fixed

File: message/tests.py
Comment (Danielsio) [reply message/tests.py:1]: @OhadSeg Thanks, good to know.

**Review by amitKaplansky — APPROVED** [review]
Looks great to me, good job (:

**Comment by IdanZimi** [comment]
After rebasing up to the most recent merge (GameEvent Entity).
I think we are ready. @yftacherzog @Yuval-Vino @OhadSeg

**Review by yftacherzog — APPROVED** [review]

File: message/tests.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+from game_event.models import GameEvent
+from court.models import Court
+from message.models import Message
+from decimal import Decimal
+from django.core.exceptions import ValidationError
+from datetime import datetime
+import pytest
+
+TEST_ID = 2
+TEST_TIME = datetime.today()
```
Comment (yftacherzog) [inline message/tests.py:12]: I noticed @MaayanMashhadi just created a PR for moving away from using this library. Can you check if this is relevant here as well?

Comment (IdanZimi) [reply message/tests.py:12]: yes, we get warnings as well. @yftacherzog

Comment (yftacherzog) [reply message/tests.py:12]: Do you want to address it before this is merged?

Comment (IdanZimi) [reply message/tests.py:12]: @MaayanMashhadi's PR will make us change some of the tests, because we are using datetime library there.

Comment (IdanZimi) [reply message/tests.py:12]: > Do you want to address it before this is merged?

yes. handling the issue right now.

**Review by Yuval-Vino — APPROVED** [review]
Well done guys!

File: message/tests.py (hunk 9 above)
Comment (IdanZimi) [reply message/tests.py]: @yftacherzog oh ok, got it. 
I changed the same issue at our `message/tests.py` .
